strange error SQL in log

### What did you do?

[2024/03/21 05:32:52.256 +00:00] [ERROR] [log.go:98] ["trace log"] [component=gorm] [elapsed=3.570094ms] [sql="DELETE FROM `xxx_meta_kvs` WHERE job_id = '341651-dump' AND meta_key like '/data-flow/dm/dropped-columns/'%"] [affected-rows=0] [error="Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '%' at line 1"]

[2024/03/21 05:32:52.256 +00:00] [ERROR] [dm_jobmaster.go:348] ["failed to clear ddl metadata"] [job_id=341651-dump] [error="[DFLOW:ErrMetaOpFail]meta operation fail: Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '%' at line 1"]

### What did you expect to see?

_No response_

### What did you see instead?

see above

### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

```console
(paste Dataflow Engine version here)
```

--- a/engine/pkg/meta/internal/sqlkv/sql_impl.go
+++ b/engine/pkg/meta/internal/sqlkv/sql_impl.go
@@ -37,7 +37,7 @@ import (
 const (
 	WhereClauseWithJobID     = "job_id = ?"
 	WhereClauseWithKeyRange  = "meta_key >= ? AND meta_key < ?"
-	WhereClauseWithKeyPrefix = "meta_key like ?%"
+	WhereClauseWithKeyPrefix = "meta_key like ?"
 	WhereClauseWithFromKey   = "meta_key >= ?"
 	WhereClauseWithKey       = "meta_key = ?"
 )
@@ -241,7 +241,10 @@ func (c *sqlKVClientImpl) doDelete(ctx context.Context, db *gorm.DB, op *metaMod
 		err = db.Where(WhereClauseWithKeyRange, key,
 			op.RangeBytes()).Delete(&sqlkvModel.MetaKV{}).Error
 	case op.IsOptsWithPrefix():
-		err = db.Where(WhereClauseWithKeyPrefix, key).Delete(&sqlkvModel.MetaKV{}).Error
+		keyPrefix := make([]byte, len(key)+1)
+		copy(keyPrefix, key)
+		keyPrefix[len(key)] = '%'
+		err = db.Where(WhereClauseWithKeyPrefix, keyPrefix).Delete(&sqlkvModel.MetaKV{}).Error
 	case op.IsOptsWithFromKey():
 		err = db.Where(WhereClauseWithFromKey, key).Delete(&sqlkvModel.MetaKV{}).Error
 	default:

--- a/engine/pkg/meta/internal/sqlkv/sql_impl.go
+++ b/engine/pkg/meta/internal/sqlkv/sql_impl.go
@@ -180,7 +180,10 @@ func (c *sqlKVClientImpl) doGet(ctx context.Context, db *gorm.DB, op *metaModel.
 	case op.IsOptsWithRange():
 		err = db.Where(WhereClauseWithKeyRange, key, op.RangeBytes()).Find(&metaKvs).Error
 	case op.IsOptsWithPrefix():
-		err = db.Where(WhereClauseWithKeyPrefix, key).Find(&metaKvs).Error
+		keyPrefix := make([]byte, len(key)+1)
+		copy(keyPrefix, key)
+		keyPrefix[len(key)] = '%'
+		err = db.Where(WhereClauseWithKeyPrefix, keyPrefix).Find(&metaKvs).Error
 	case op.IsOptsWithFromKey():
 		err = db.Where(WhereClauseWithFromKey, key).Find(&metaKvs).Error
 	default:

--- a/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
+++ b/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
@@ -240,8 +240,8 @@ func TestGet(t *testing.T) {
 				},
 			},
 			mockExpectResFn: func(mock sqlmock.Sqlmock) {
-				mock.ExpectQuery(regexp.QuoteMeta("SELECT * FROM `fakeTable` WHERE job_id = ? AND meta_key like ?%")).
-					WithArgs(fakeJob, []byte("key0")).
+				mock.ExpectQuery(regexp.QuoteMeta("SELECT * FROM `fakeTable` WHERE job_id = ? AND meta_key like ?")).
+					WithArgs(fakeJob, []byte("key0%")).
 					WillReturnRows(sqlmock.NewRows([]string{"meta_key", "meta_value"}).
 						AddRow("key0", "value0"))
 			},
@@ -319,8 +319,8 @@ func TestDelete(t *testing.T) {
 				Header: &metaModel.ResponseHeader{},
 			},
 			mockExpectResFn: func(mock sqlmock.Sqlmock) {
-				mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `fakeTable` WHERE job_id = ? AND meta_key like ?%")).
-					WithArgs(fakeJob, []byte("key0")).
+				mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `fakeTable` WHERE job_id = ? AND meta_key like ?")).
+					WithArgs(fakeJob, []byte("key0%")).
 					WillReturnResult(sqlmock.NewResult(1, 1))
 			},
 		},

--- a/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
+++ b/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
@@ -356,8 +356,8 @@ func TestTxn(t *testing.T) {
 		"`meta_key`=VALUES(`meta_key`),`meta_value`=VALUES(`meta_value`),`job_id`=VALUES(`job_id`)")).
 		WithArgs(anyT, anyT, []byte("key1"), []byte("value1"), fakeJob, anyT).
 		WillReturnResult(sqlmock.NewResult(1, 1))
-	mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `fakeTable` WHERE job_id = ? AND meta_key like ?%")).
-		WithArgs(fakeJob, []byte("key2")).
+	mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `fakeTable` WHERE job_id = ? AND meta_key like ?")).
+		WithArgs(fakeJob, []byte("key2%")).
 		WillReturnResult(sqlmock.NewResult(1, 1))
 	mock.ExpectCommit()
 
@@ -407,6 +407,7 @@ func testInner(t *testing.T, m sqlmock.Sqlmock, cli *sqlKVClientImpl, c tCase) {
 			require.Equal(t, c.output, result[0].Interface())
 		}
 	}
+	require.NoError(t, m.ExpectationsWereMet())
 }
 
 func TestSQLImplWithoutNamespace(t *testing.T) {
@@ -501,4 +502,5 @@ func TestInitializeError(t *testing.T) {
 		WillReturnError(errors.New("other error"))
 	_, err = NewSQLKVClientImpl(db, defaultTestStoreType, "test", "")
 	require.Regexp(t, "other error", err.Error())
+	require.NoError(t, mock.ExpectationsWereMet())
 }

--- a/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
+++ b/engine/pkg/meta/internal/sqlkv/sql_impl_test.go
@@ -448,8 +448,8 @@ func TestSQLImplWithoutNamespace(t *testing.T) {
 		"`meta_key`=VALUES(`meta_key`),`meta_value`=VALUES(`meta_value`),`job_id`=VALUES(`job_id`)")).
 		WithArgs(anyT, anyT, []byte("key1"), []byte("value1"), "", anyT).
 		WillReturnResult(sqlmock.NewResult(1, 1))
-	mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `meta_kvs` WHERE job_id = ? AND meta_key like ?%")).
-		WithArgs("", []byte("key2")).
+	mock.ExpectExec(regexp.QuoteMeta("DELETE FROM `meta_kvs` WHERE job_id = ? AND meta_key like ?")).
+		WithArgs("", []byte("key2%")).
 		WillReturnResult(sqlmock.NewResult(1, 1))
 	mock.ExpectCommit()
 	txn := cli.Txn(context.Background())
@@ -457,6 +457,7 @@ func TestSQLImplWithoutNamespace(t *testing.T) {
 	txn.Do(metaModel.OpPut("key1", "value1"))
 	txn.Do(metaModel.OpDelete("key2", metaModel.WithPrefix()))
 	txn.Commit()
+	require.NoError(t, mock.ExpectationsWereMet())
 }
 
 func TestInitializeError(t *testing.T) {

--- a/engine/test/integration_tests/dm_lightning_checkpoint/run.sh
+++ b/engine/test/integration_tests/dm_lightning_checkpoint/run.sh
@@ -22,9 +22,7 @@ function run() {
 	# create job & wait for job finished
 	job_id=$(create_job "DM" "$CUR_DIR/conf/job.yaml" "checkpoint")
 	job_id2=$(create_job "DM" "$CUR_DIR/conf/job2.yaml" "checkpoint2")
-	exec_with_retry --count 30 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id/status\" | tee /dev/stderr | jq -e '.task_status.\"mysql-01\".status.unit == \"DMLoadTask\"'" &
-	exec_with_retry --count 30 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id2/status\" | tee /dev/stderr | jq -e '.task_status.\"mysql-02\".status.unit == \"DMLoadTask\"'"
-	exec_with_retry --count 100 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id\" | tee /dev/stderr | jq -e '.state == \"Finished\"'" &
+	exec_with_retry --count 100 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id\" | tee /dev/stderr | jq -e '.state == \"Finished\"'"
 	exec_with_retry --count 100 "curl \"http://127.0.0.1:10245/api/v1/jobs/$job_id2\" | tee /dev/stderr | jq -e '.state == \"Finished\"'"
 	# clean the lighting checkpoint in downstream after job is finished
 	exec_with_retry --count 30 '! run_sql --port 4000 "show databases;" | grep -q "checkpoint"'

